File: modules/module-runtime/frontend/src/manifest.ts
export type ModuleStatus = "active" | "experimental" | "deprecated";
export type ModuleAvailability = "enabled" | "disabled" | "blocked";

export interface ModuleManifest {
  readonly schema_version: 1;
  readonly id: string;
  readonly version: string;
  readonly title: string;
  readonly description: string;
  readonly status: ModuleStatus;
  readonly feature_flag: string;
  readonly requires: {
    readonly modules: readonly string[];
    readonly integrations: readonly string[];
    readonly optional_integrations: readonly string[];
  };
  readonly contributes: {
    readonly editors: readonly string[];
    readonly commands: readonly string[];
    readonly events: readonly string[];
    readonly jobs: readonly string[];
    readonly workflows: readonly string[];
    readonly policy_gates: readonly string[];
  };
  readonly permissions: readonly string[];
  readonly entrypoints: {
    readonly frontend?: string;
    readonly backend?: string;
  };
}

export interface ModuleContribution {
  readonly manifest: ModuleManifest;
  readonly editors?: readonly unknown[];
  readonly commands?: readonly unknown[];
  readonly events?: readonly unknown[];
  readonly jobs?: readonly unknown[];
}

export interface ModuleRuntimeState {
  readonly moduleId: string;
  readonly featureFlag: string;
  readonly availability: ModuleAvailability;
  readonly missingRequiredIntegrations: readonly string[];
  readonly missingOptionalIntegrations: readonly string[];
  readonly blockingDependencies: readonly string[];
  readonly message: string;
}

export function resolveFrontendModuleStates(
  manifests: readonly ModuleManifest[],
  featureFlags: Readonly<Record<string, boolean>> = {},
  availableIntegrations: ReadonlySet<string> = new Set(),
): ReadonlyMap<string, ModuleRuntimeState> {
  const knownFlags = new Set(manifests.map((manifest) => manifest.feature_flag));
  const unknownFlags = Object.keys(featureFlags).filter((flag) => !knownFlags.has(flag));
  if (unknownFlags.length > 0) {
    throw new Error(`UNKNOWN_FEATURE_FLAGS: ${unknownFlags.sort().join(", ")}`);
  }

  const states = new Map<string, ModuleRuntimeState>();
  for (const manifest of manifests) {
    const missingRequiredIntegrations = manifest.requires.integrations
      .filter((integration) => !availableIntegrations.has(integration))
      .toSorted();
    const missingOptionalIntegrations = manifest.requires.optional_integrations
      .filter((integration) => !availableIntegrations.has(integration))
      .toSorted();
    const blockingDependencies = manifest.requires.modules
      .filter((dependency) => states.get(dependency)?.availability !== "enabled")
      .toSorted();

    let availability: ModuleAvailability;
    let message: string;
    if (featureFlags[manifest.feature_flag] === false) {
      availability = "disabled";
      message = "功能已由特性开关关闭。";
    } else if (blockingDependencies.length > 0) {
      availability = "blocked";
      message = `依赖模块不可用：${blockingDependencies.join("、")}`;
    } else if (missingRequiredIntegrations.length > 0) {
      availability = "blocked";
      message = `缺少必需集成：${missingRequiredIntegrations.join("、")}`;
    } else {
      availability = "enabled";
      message =
        missingOptionalIntegrations.length > 0
          ? `已启用；可选集成不可用：${missingOptionalIntegrations.join("、")}`
          : "已启用。";
    }
    states.set(manifest.id, {
      moduleId: manifest.id,
      featureFlag: manifest.feature_flag,
      availability,
      missingRequiredIntegrations,
      missingOptionalIntegrations,
      blockingDependencies,
      message,
    });
  }
  return states;
}

File: modules/module-runtime/backend/src/module_runtime/manifest.py
"""Typed source-of-truth model for ``modules/*/module.yaml``."""

from __future__ import annotations

from enum import Enum
from typing import Annotated, List, Optional

from pydantic import BaseModel, ConfigDict, Field, StringConstraints, model_validator


MODULE_ID_PATTERN = r"^[a-z][a-z0-9]*(?:-[a-z0-9]+)*$"
FEATURE_FLAG_PATTERN = r"^[a-z][a-z0-9]*(?:_[a-z0-9]+)*$"
CONTRIBUTION_ID_PATTERN = r"^[a-z][a-z0-9]*(?:[._-][a-z0-9]+)+$"
EVENT_REFERENCE_PATTERN = (
    r"^[a-z][a-z0-9]*(?:\.[a-z][a-z0-9_]*)+@[1-9][0-9]*$"
)
PERMISSION_PATTERN = r"^[a-z][a-z0-9-]*(?::[a-z][a-z0-9_-]*)+$"
FRONTEND_ENTRYPOINT_PATTERN = (
    r"^\./(?!(?:.*?/)?\.\.(?:/|$))(?:[A-Za-z0-9_-]+/)*[A-Za-z0-9_-]+\.tsx?$"
)
BACKEND_ENTRYPOINT_PATTERN = r"^[a-z][a-z0-9_]*(?:\.[a-z][a-z0-9_]*)*$"

ModuleIdValue = Annotated[str, StringConstraints(pattern=MODULE_ID_PATTERN)]
ContributionIdValue = Annotated[str, StringConstraints(pattern=CONTRIBUTION_ID_PATTERN)]
EventReferenceValue = Annotated[str, StringConstraints(pattern=EVENT_REFERENCE_PATTERN)]
PermissionValue = Annotated[str, StringConstraints(pattern=PERMISSION_PATTERN)]


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)


class ModuleStatus(str, Enum):
    ACTIVE = "active"
    EXPERIMENTAL = "experimental"
    DEPRECATED = "deprecated"


class ModuleRequirements(StrictModel):
    modules: List[ModuleIdValue]
    integrations: List[ModuleIdValue]
    optional_integrations: List[ModuleIdValue]

    @model_validator(mode="after")
    def validate_lists(self) -> "ModuleRequirements":
        fields = {
            "modules": self.modules,
            "integrations": self.integrations,
            "optional_integrations": self.optional_integrations,
        }
        for name, values in fields.items():
            if len(values) != len(set(values)):
                raise ValueError(f"{name} cannot contain duplicates")
        overlap = set(self.integrations) & set(self.optional_integrations)
        if overlap:
            raise ValueError(
                "integrations and optional_integrations overlap: "
                + ", ".join(sorted(overlap))
            )
        return self


class ModuleContributions(StrictModel):
    editors: List[ContributionIdValue]
    commands: List[ContributionIdValue]
    events: List[EventReferenceValue]
    jobs: List[ContributionIdValue]
    workflows: List[ContributionIdValue]
    policy_gates: List[ContributionIdValue]

    @model_validator(mode="after")
    def validate_contributions(self) -> "ModuleContributions":
        for name in ("editors", "commands", "jobs", "workflows", "policy_gates"):
            values = getattr(self, name)
            if len(values) != len(set(values)):
                raise ValueError(f"{name} cannot contain duplicates")
        if len(self.events) != len(set(self.events)):
            raise ValueError("events cannot contain duplicates")
        return self


class ModuleEntrypoints(StrictModel):
    frontend: Optional[str] = Field(
        default=None, json_schema_extra={"pattern": FRONTEND_ENTRYPOINT_PATTERN}
    )
    backend: Optional[str] = Field(default=None, pattern=BACKEND_ENTRYPOINT_PATTERN)

    @model_validator(mode="after")
    def require_surface(self) -> "ModuleEntrypoints":
        if self.frontend is None and self.backend is None:
            raise ValueError("at least one public entrypoint is required")
        if self.frontend is not None and not _fullmatch(
            FRONTEND_ENTRYPOINT_PATTERN, self.frontend
        ):
            raise ValueError("frontend entrypoint must be a safe module-local .ts/.tsx path")
        return self


class ModuleManifest(StrictModel):
    schema_version: int = Field(ge=1, le=1)
    id: str = Field(pattern=MODULE_ID_PATTERN)
    version: str = Field(pattern=r"^(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)\.(0|[1-9][0-9]*)$")
    title: str = Field(min_length=1, max_length=120)
    description: str = Field(min_length=1, max_length=500)
    status: ModuleStatus
    feature_flag: str = Field(pattern=FEATURE_FLAG_PATTERN)
    requires: ModuleRequirements
    contributes: ModuleContributions
    permissions: List[PermissionValue]
    entrypoints: ModuleEntrypoints

    @model_validator(mode="after")
    def validate_manifest(self) -> "ModuleManifest":
        if self.id in self.requires.modules:
            raise ValueError("module cannot require itself")
        if len(self.permissions) != len(set(self.permissions)):
            raise ValueError("permissions cannot contain duplicates")
        return self


def _fullmatch(pattern: str, value: str) -> bool:
    import re

    return re.fullmatch(pattern, value) is not None
